### PR TITLE
[WIP] Postgresql errors

### PR DESCRIPTION
**Pull request by julsbreakdown**
#3233

### CODE REVIEW

**Comment by sbrunner** [comment]
You target 2.2 or master?

**Comment by julsbreakdown** [comment]
2.2

**Comment by julsbreakdown** [comment]
Are there any tests done for feature editing?

**Comment by sbrunner** [comment]
Then I just change the base :-)

**Comment by sbrunner** [comment]
For me it looks strange to have the error type in the key, shouldn't we have something like ?:
```json
{
  "error_type": ...,
  "message": ...
}
```

and this documentation should be updated: https://github.com/camptocamp/c2cgeoportal/blob/2.2/doc/developer/webservices.rst#mapfish-protocol

**Comment by sbrunner** [comment]
The tests should be here:https://github.com/camptocamp/c2cgeoportal/blob/2.2/c2cgeoportal/tests/functional/test_layers.py

**Comment by julsbreakdown** [comment]
Ok I will apply the changes + docs. thank you

File: doc/developer/webservices.rst [hunk 1]
```diff
@@ -408,6 +408,66 @@ Result
...
+                              "good": true,
+                              "name": "nom",
+                              "internal_id": null,
+                              "short_name": "court",
+                              "height": null,
+                              "short_name3": "R",
+                              "short_name2": "2"
+                              }
+             }]
+}
```
Comment (sbrunner) [inline doc/developer/webservices.rst]: Can you format it as standard JSON not this ugly one :-(

File: doc/developer/webservices.rst [hunk 2]
```diff
@@ -408,6 +408,66 @@ Result
...
+
+Error :
+.. code:: json
+
+ {"message" : 'error description, "error_type", 'type of error'}
```
Comment (sbrunner) [inline doc/developer/webservices.rst]: single quote is forbinden in json =>
```json
{
    "error_type": "type of error",
    "message" : "error description"
}
```

File: doc/developer/webservices.rst [hunk 3]
```diff
@@ -408,6 +408,66 @@ Result
         }, ...]
     }
 
+Update feature
+------------------
+
+URL: ``.../layers/<layer_name>/<layer_id>/<feature_id>``
+
+Success : geojson
```
Comment (sbrunner) [inline doc/developer/webservices.rst]: No space before : in english...

**Comment by julsbreakdown** [comment]
Ok documentation is up to date. I still can't launch the tests though..

File: doc/developer/webservices.rst [hunk 4]
```diff
@@ -408,6 +408,97 @@ Result
...
+------------------
+
+URL: ``.../layers/<layer_name>/<layer_id>/<feature_id>``
+
+Success:
+
+.. code:: json
+
+   {
+       "type":"FeatureCollection",
```
Comment (sbrunner) [inline doc/developer/webservices.rst]: add space after `:` :-)

File: doc/developer/webservices.rst [hunk 5]
```diff
@@ -408,6 +408,97 @@ Result
...
+             }
+          }
+       ]
+   }
+
+Error :
+
+.. code:: json
+
+ {"message" : "error description", "error_type", "type of error"}
```
Comment (sbrunner) [inline doc/developer/webservices.rst]: error on red :-)
more readable on 4 lines :-)

**Comment by julsbreakdown** [comment]
@sbrunner 
I can not find a way to debug see https://github.com/camptocamp/c2cgeoportal/pull/3238/files#diff-0931bbce880d0190da443d80aec059c3R578

File: c2cgeoportal/views/layers.py [hunk 6]
```diff
@@ -324,7 +337,20 @@ def check_geometry(r, feature, o):
...
+        except exc.IntegrityError as e:
+            self.request.response.status_int = 400
+            return {"error_type": "integrity_error",
+                    "message": str(e.orig.diag.message_primary)}
+        except exc.ProgrammingError as e:
+            self.request.response.status_int = 400
+            return {"error_type": "programming_error",
+                    "message": str(e.orig.diag.message_primary)}
+        except exc.DatabaseError as e:
+            self.request.response_status_int = 400
```
Comment (sbrunner) [inline c2cgeoportal/views/layers.py]: => self.request.response.status_int

File: c2cgeoportal/views/layers.py [hunk 7]
```diff
@@ -272,7 +272,20 @@ def check_geometry(r, feature, o):
...
```
Comment (sbrunner) [inline c2cgeoportal/views/layers.py]: => self.request.response.status_int

**Comment by sbrunner** [comment]
Isn't the new comment?

**Comment by julsbreakdown** [comment]
sorry I do not understand your last comment =(

**Comment by sbrunner** [comment]
? me neither :-P

**Comment by julsbreakdown** [comment]
=)
I have added a constraint : "email" type.
I have also created an extra test that adds a non valid email.
This should return a 400 error ('integrity error')  but instead I get 201.
(See the link in my previous comment)

I am not able to debug (Enable to use pdb/ipdb in the docker container).
=> Maybe the constraint isn't created in the table
=> Can you have a look?

Thank you

**Comment by sbrunner** [comment]
Run one test ```./docker-compose-run nosetests --where /src/c2cgeoportal/tests/functional/test_layers.py:TestLayers.test_update_validation_fails_constraint```

After the test run (docker compose still running):
docker exec -ti c2cgeoportal_db_1 bash
psql -h localhost -U www-data -W geomapfish_test
select * from table_1;
```
 id | child_id | name |    email     | last_update_user | last_update_date |                        geom                        
----+----------+------+--------------+------------------+------------------+----------------------------------------------------
  2 |        2 | bar  |              |                  |                  | 01010000201555000000000000000018400000000000004740
  1 |          | foo  | novalidemail |                  |                  | 01010000201555000000000000000014400000000000804640
```
\d table_1
```
      Column      |            Type             |                      Modifiers                       
------------------+-----------------------------+------------------------------------------------------
 id               | integer                     | not null default nextval('table_1_id_seq'::regclass)
 child_id         | integer                     | 
 name             | character varying           | 
 email            | character varying           | 
 last_update_user | character varying           | 
 last_update_date | timestamp without time zone | 
 geom             | geometry(Geometry,21781)    | 
Indexes:
    "table_1_pkey" PRIMARY KEY, btree (id)
    "idx_table_1_geom" gist (geom)
Foreign-key constraints:
    "table_1_child_id_fkey" FOREIGN KEY (child_id) REFERENCES table_1_child(id)
```
=> no constraints...

File: c2cgeoportal/tests/functional/test_layers.py [hunk 8]
```diff
@@ -138,7 +136,8 @@ def _create_layer(
             Column("id", types.Integer, primary_key=True),
             Column("child_id", types.Integer,
                    ForeignKey("public.{0!s}_child.id".format(tablename))),
-            Column("name", types.Unicode),
+            Column("name", types.Unicode, CheckConstraint("email ~* '^[A-Za-z0-9._%-]+@[A-Za-z0-9.-]+[.][A-Za-z]+$'",
+                                   name="proper_email")),
```
Comment (sbrunner) [inline c2cgeoportal/tests/functional/test_layers.py]: email check constraint on name?

**Comment by julsbreakdown** [comment]
-_-'

On Fri, Nov 17, 2017 at 12:39 PM, Stéphane Brunner <notifications@github.com

**Pull request merged**